src/ratelimit/RateLimitController.ts

```diff
@@ -0,0 +1,187 @@
+  RateLimitStateChange<ApiType>,
+  never,
+  never
+>;
+
+const metadata = {
+  requests: { persist: false, anonymous: false },
+};
+
+/**
+ * Controller that handles showing notifications to the user and rate limiting origins
+ */
+export class RateLimitController<ApiType extends string> extends BaseController<
+  typeof name,
+  RateLimitState<ApiType>,
+  RateLimitMessenger<ApiType>
+> {
+  private implementations;
+
+  private rateLimitTimeout;
+
+  private rateLimitCount;
+
+  /**
+   * Creates a RateLimitController instance.
+   *
+   * @param options - Constructor options.
+   * @param options.messenger - A reference to the messaging system.
+   * @param options.state - Initial state to set on this controller.
+   * @param options.implementations - Mapping from API type to API implementation
+   * @param options.rateLimitTimeout - The time window in which the rate limit is applied
+   * @param options.rateLimitCount - The amount of notifications an origin can show in the rate limit time window
+   */
+  constructor({
+    // @todo Pick some sane defaults for this
+    rateLimitTimeout = 5000,
+    rateLimitCount = 3,
+    messenger,
+    state,
+    implementations,
+  }: {
+    rateLimitTimeout?: number;
+    rateLimitCount?: number;
+    messenger: RateLimitMessenger<ApiType>;
+    state?: Partial<RateLimitState<ApiType>>;
+    implementations: Record<ApiType, (...args: unknown[]) => unknown>;
+  }) {
+    const defaultState = {
+      requests: Object.keys(implementations).reduce(
+        (acc, key) => ({ ...acc, [key]: {} }),
+        {} as Record<ApiType, Record<string, number>>,
+      ),
+    };
+    super({
+      name,
+      metadata,
+      messenger,
+      state: { ...defaultState, ...state },
+    });
+    this.implementations = implementations;
+    this.rateLimitTimeout = rateLimitTimeout;
+    this.rateLimitCount = rateLimitCount;
+
+    this.messagingSystem.registerActionHandler(
+      `${name}:call` as const,
+      (origin: string, args: CallArgs<ApiType>) => this.call(origin, args),
+    );
+  }
+
+  /**
+   * Shows a notification if origin is not rate-limited.
+   *
+   * @param origin - The origin trying to send a notification
+   * @param _args - Notification arguments, containing the notification message etc.
+   * @returns False if rate-limited, true if not
+   */
+  call(origin: string, _args: CallArgs<ApiType>) {
+    const { type, args } = _args;
+    if (this._isRateLimited(type, origin)) {
+      return false;
+    }
+    this._recordRequest(type, origin);
+
+    const implementation = this.implementations[type];
+
+    if (!implementation) {
+      throw new Error('Invalid api type');
+    }
+
+    implementation(...(args ?? []));
+
+    return true;
+  }
+
+  /**
+   * Checks whether a given origin is rate limited for a given API.
+   *
+   * @param api - The API the origin is trying to access
+   * @param origin - The origin trying to access the API
+   * @returns True if rate-limited
```

Comment:
```suggestion
   * Checks whether an origin is rate limited for the a specific API.
   *
   * @param api - The API the origin is trying to access.
   * @param origin - The origin trying to access the API.
   * @returns `true` if rate-limited, and `false` otherwise.
```